taichi/common/serialization.h

```diff
@@ -54,71 +56,93 @@ using is_unit = typename std::is_base_of<Unit, remove_cvref_t<T>>;
 using is_unit_t = typename is_unit<T>::type;
+
 }  // namespace type
 
-#define TI_IO_DECL_INST                               \
-  void binary_io(BinaryOutputSerializer &ser) const { \
-    ser(*this);                                       \
-  }                                                   \
-  void binary_io(BinaryInputSerializer &ser) const {  \
-    ser(*this);                                       \
-  }
+namespace detail {
 
-#define TI_IO_DECL_INST_VIRT                                  \
-  virtual void binary_io(BinaryOutputSerializer &ser) const { \
-    ser(*this);                                               \
-  }                                                           \
-  virtual void binary_io(BinaryInputSerializer &ser) const {  \
-    ser(*this);                                               \
+template <size_t N>
+constexpr size_t count_delim(const char (&str)[N], char delim) {
+  size_t count = 1;
+  for (const char &ch : str) {
+    if (ch == delim) {
+      ++count;
+    }
   }
+  return count;
+}
 
-#define TI_IO_DECL_INST_VIRT_OVERRIDE                                  \
-  virtual void binary_io(BinaryOutputSerializer &ser) const override { \
-    ser(*this);                                                        \
-  }                                                                    \
-  virtual void binary_io(BinaryInputSerializer &ser) const override {  \
-    ser(*this);                                                        \
+template <size_t DelimN>
+struct StrDelimSplitter {
+  template <size_t StrsN>
+  static constexpr std::array<std::string_view, DelimN> make(
+      const char (&str)[StrsN],
+      char delim) {
+    std::array<std::string_view, DelimN> res;
+    const char *head = &(str[0]);
+    size_t si = 0;
+    size_t cur_head_i = 0;
+    size_t ri = 0;
+    while (si < StrsN) {
+      if (str[si] != delim) {
+        si += 1;
+      } else {
+        res[ri] = {head, (si - cur_head_i)};
+        ++ri;
+        si += 2;  // skip ", "
+        cur_head_i = si;
+        head = &(str[cur_head_i]);
+      }
+    }
+    // `StrsN - 1` because the last char is '\0'.
+    res[ri] = {head, (StrsN - 1 - cur_head_i)};
+    return res;
   }
+};
 
-#define TI_IO_DECL      \
-  TI_IO_DECL_INST       \
-  template <typename S> \
-  void io(S &serializer) const
+template <typename SER, size_t N, typename T>
+void serialize_kv_impl(SER &ser,
+                       const std::array<std::string_view, N> &keys,
+                       T &&val) {
+  std::string key{keys[N - 1]};
+  ser(key.c_str(), val);
+}
 
-#define TI_IO_DECL_VIRT \
-  TI_IO_DECL_INST_VIRT  \
-  template <typename S> \
-  void io(S &serializer) const
+template <typename SER, size_t N, typename T, typename... Args>
+void serialize_kv_impl(SER &ser,
+                       const std::array<std::string_view, N> &keys,
+                       T &&head,
+                       Args &&... rest) {
+  constexpr auto i = (N - sizeof...(Args));
```

Comment:
Looks like this is reversing the keys -- why not set `i` as a parameter or a template parameter? Or, why do the resulting strings look not reversed?

Reply:
> Looks like this is reversing the keys 
Not really, `sizeof...(Args)` decrements after each call, so `i` is actually incrementing.